src/Analysis/Ast/Impl/Analyzer/Evaluation/ExpressionEval.Generics.cs

```diff
@@ -26,97 +27,72 @@ namespace Microsoft.Python.Analysis.Analyzer.Evaluation {
     internal sealed partial class ExpressionEval {
         /// <summary>
         /// Creates specific type from expression that involves generic type
-        /// and the specific type arguments.
+        /// and the specific type arguments, such as Generic[T] or constructor
+        /// of a generic class.
         /// </summary>
-        /// <example>
-        ///     x = List[T]
-        /// </example>
-        /// <example>
-        ///     T = TypeVar('T', Exception)
-        ///     class A(Generic[T]): ...
-        ///     x = A(TypeError)
-        /// </example>
         private IMember GetValueFromGeneric(IMember target, Expression expr) {
             if (!(target is PythonClassType c && c.IsGeneric()) && !(target is IGenericType)) {
                 return null;
             }
-            // Evaluate index to check if the result is generic parameter.
-            // If it is, then this is a declaration expression such as Generic[T]
-            // rather than specific type instantiation as in List[str].
-
-            IPythonType[] specificTypes;
-            var returnInstance = false;
-            switch (expr) {
-                // Indexing returns type as from A[int]
-                case IndexExpression indexExpr:
-                    // Generic[T1, T2, ...] or A[type]()
-                    var indices = EvaluateIndex(indexExpr);
-                    // See which ones are generic parameters as defined by TypeVar() 
-                    // and which are specific types. Normally there should not be a mix.
-                    var genericTypeArgs = indices.OfType<IGenericTypeParameter>().ToArray();
-                    specificTypes = indices.Where(i => !(i is IGenericTypeParameter)).OfType<IPythonType>().ToArray();
-
-                    if (specificTypes.Length == 0 && genericTypeArgs.Length > 0) {
-                        // The expression is still generic. For example, generic return
-                        // annotation of a class method, such as 'def func(self) -> A[_E]: ...'.
-                        // Leave it alone, we don't want resolve generic with generic.
-                        return null;
-                    }
 
-                    if (genericTypeArgs.Length > 0 && genericTypeArgs.Length != indices.Count) {
-                        // TODO: report that some type arguments are not declared with TypeVar.
-                    }
+            using (OpenScope(target.GetPythonType()?.DeclaringModule, GetScope(target), out _)) {
```

Comment:
`OpenScope` changes `Scope CurrentScope { get; private set; }`. Should it be defined as `AsyncLocal<Scope>`?

Reply:
During analysis it is not a problem since there is no concurrency anymore. But there is indeed a bit of a problem with it indeed when several requests for tooltips may come interleaved. I've seen AF couple of times when trying to close scope that was not open. AF is disabled for now since nothing bad happens really. I could make it `AsyncLocal` if that would help, but I am not entirely sure.

Reply:
Yeah, if there are no concurrent requests, `AsyncLocal` won't help.